File: modules/browser/src/main/java/org/gjt/jclasslib/browser/detail/TypeAnnotationDetailPane.java
/*
 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public
 License as published by the Free Software Foundation; either
 version 2 of the license, or (at your option) any later version.
 */

/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/
package org.gjt.jclasslib.browser.detail;

import org.gjt.jclasslib.browser.BrowserServices;
import org.gjt.jclasslib.browser.BrowserTreeNode;
import org.gjt.jclasslib.browser.NodeType;
import org.gjt.jclasslib.structures.AttributeInfo;
import org.gjt.jclasslib.structures.MethodInfo;
import org.gjt.jclasslib.structures.attributes.*;
import org.gjt.jclasslib.util.ExtendedJLabel;
import org.gjt.jclasslib.util.HtmlDisplayTextArea;

import javax.swing.*;
import javax.swing.event.HyperlinkEvent;
import javax.swing.event.HyperlinkListener;
import javax.swing.tree.TreePath;

/**
 * Class for showing a type annotation node.
 */
public class TypeAnnotationDetailPane extends FixedListDetailPane {

    private ExtendedJLabel lblType;
    private HtmlDisplayTextArea txtInfo;
    private HtmlDisplayTextArea txtType;
    private ExtendedJLabel lblTypePath;

    public TypeAnnotationDetailPane(BrowserServices services) {
        super(services);
    }

    protected void addLabels() {
        addDetailPaneEntry(normalLabel("Target Type:"),
            lblType = highlightLabel());

        addDetailPaneEntry(normalLabel("Target Info:"),
            txtInfo = highlightTextArea());

        addDetailPaneEntry(lblTypePath = normalLabel("Type path:"),
            txtType = highlightTextArea());

        txtInfo.addHyperlinkListener(new HyperlinkListener() {
            @Override
            public void hyperlinkUpdate(HyperlinkEvent e) {
                if (e.getEventType() == HyperlinkEvent.EventType.ACTIVATED) {
                    String description = e.getDescription();
                    handleLink(description);
                }
            }
        });

    }

    public void show(TreePath treePath) {
        TypeAnnotation typeAnnotation = (TypeAnnotation)((BrowserTreeNode)treePath.getLastPathComponent()).getElement();

        lblType.setText(typeAnnotation.getTargetType().toString());
        txtInfo.setText(typeAnnotation.getTargetInfo().getVerbose().replace("\n", "<br>"));
        String typePathVerbose = getTypePathVerbose(typeAnnotation.getTypePathEntries());
        boolean pathAvailable = typePathVerbose != null;
        if (pathAvailable) {
            txtType.setText(typePathVerbose);
        }
        lblTypePath.setVisible(pathAvailable);
        txtType.setVisible(pathAvailable);

        super.show(treePath);
    }

    private String getTypePathVerbose(TypePathEntry[] typePathEntries) {
        if (typePathEntries == null || typePathEntries.length == 0) {
            return null;
        }
        StringBuilder buffer = new StringBuilder();
        for (TypePathEntry typePathEntry : typePathEntries) {
            buffer.append(typePathEntry.getTypePathKind());
            buffer.append(", argument index ").append(typePathEntry.getTypeArgumentIndex());
        }
        return buffer.toString();
    }

    private void handleLink(String description) {
        char type = description.charAt(0);
        int index = Integer.parseInt(description.substring(1));
        switch (type) {
            case 'E':
                handleExceptionsLink(index);
                break;
            case 'L':
                handleLocalVarLink(index);
                break;
            case 'I':
                handleInterfaceLink(index);
                break;
            default:
                throw new IllegalArgumentException("Invalid link type " + type);
        }
    }

    private void handleExceptionsLink(int index) {
        JTree tree = getServices().getBrowserComponent().getTreePane().getTree();
        TreePath parentPath = findParentNode(tree, MethodInfo.class);
        TreePath path = findAttributeChildNode(parentPath, ExceptionsAttribute.class);

        handleListLink(index, path, ExceptionsAttribute.class);
    }

    private void handleLocalVarLink(int index) {
        JTree tree = getServices().getBrowserComponent().getTreePane().getTree();
        TreePath parentPath = findParentNode(tree, CodeAttribute.class);
        TreePath path = findAttributeChildNode(parentPath, LocalVariableTableAttribute.class);
        LocalVariableTableAttribute attribute = (LocalVariableTableAttribute)((BrowserTreeNode)path.getLastPathComponent()).getElement();

        handleListLink(getLinkIndex(index, attribute), path, LocalVariableTableAttribute.class);
    }

    private int getLinkIndex(int index, LocalVariableTableAttribute attribute) {
        LocalVariableEntry[] localVariableTable = attribute.getLocalVariableEntries();
        for (int i = 0; i < localVariableTable.length; i++) {
            LocalVariableEntry entry = localVariableTable[i];
            if (entry.getIndex() == index) {
                return i;
            }
        }
        throw new IllegalArgumentException("index " + index + " not found in local variable table");
    }

    private void handleInterfaceLink(int index) {
        TreePath interfacesPath = getServices().getBrowserComponent().getTreePane().getPathForCategory(NodeType.INTERFACE);
        BrowserTreeNode interfacesNode = (BrowserTreeNode)interfacesPath.getLastPathComponent();
        if (index >= interfacesNode.getChildCount()) {
            throw new IllegalArgumentException("Invalid interface index " + index);
        }

        TreePath path = interfacesPath.pathByAddingChild(interfacesNode.getChildAt(index));
        selectPath(path);
    }

    private void handleListLink(int index, TreePath path, Class<? extends AttributeInfo> attributeClass) {
        selectPath(path);
        AttributeDetailPane detailPane = (AttributeDetailPane)getServices().getBrowserComponent().getDetailPane().getCurrentDetailPane();
        ((TableDetailPane)detailPane.getDetailPane(attributeClass)).selectIndex(index);
    }

    private void selectPath(TreePath path) {
        JTree tree = getServices().getBrowserComponent().getTreePane().getTree();
        tree.setSelectionPath(path);
        tree.scrollPathToVisible(path);
    }

    private TreePath findAttributeChildNode(TreePath path, Class<? extends AttributeInfo> attributeClass) {
        BrowserTreeNode methodNode = (BrowserTreeNode)path.getLastPathComponent();
        for (int i = 0; i < methodNode.getChildCount(); i++) {
            BrowserTreeNode attributeNode = (BrowserTreeNode)methodNode.getChildAt(i);
            if (attributeNode.getElement().getClass() == attributeClass) {
                return path.pathByAddingChild(attributeNode);
            }
        }
        throw new RuntimeException("No attribute node for class " + attributeClass + " found");
    }

    private TreePath findParentNode(JTree tree, Class elementClass) {
        TreePath path = tree.getSelectionPath();
        while (path != null) {
            BrowserTreeNode node = (BrowserTreeNode)path.getLastPathComponent();
            if (node.getElement().getClass() == elementClass) {
                break;
            } else {
                path = path.getParentPath();
            }
        }
        if (path == null) {
            throw new RuntimeException("No parent node with element class " + elementClass + " found");
        }
        return path;
    }

}

File: modules/browser/src/main/kotlin/org/gjt/jclasslib/browser/detail/TypeAnnotationDetailPane.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/
package org.gjt.jclasslib.browser.detail

import org.gjt.jclasslib.browser.BrowserServices
import org.gjt.jclasslib.browser.BrowserTreeNode
import org.gjt.jclasslib.browser.NodeType
import org.gjt.jclasslib.browser.detail.attributes.CodeAttributeDetailPane
import org.gjt.jclasslib.structures.AttributeInfo
import org.gjt.jclasslib.structures.Structure
import org.gjt.jclasslib.structures.attributes.CodeAttribute
import org.gjt.jclasslib.structures.attributes.LocalVariableTableAttribute
import org.gjt.jclasslib.structures.attributes.TypeAnnotation
import javax.swing.JTree
import javax.swing.tree.TreePath

class TypeAnnotationDetailPane(services: BrowserServices) : KeyValueDetailPane<TypeAnnotation>(TypeAnnotation::class.java, services) {
    override fun addLabels() {
        addDetail("Target Type:") { typeAnnotation -> typeAnnotation.targetType.toString() }

        addMultiLineHtmlDetail("Target Info:") { typeAnnotation ->
            typeAnnotation.targetInfo.verbose.replace("\n", "<br>")
        }.linkHandler { description ->
            handleLink(description)
        }

        addMultiLineHtmlDetail("Type path:") { typeAnnotation ->
            StringBuilder().apply {
                for (typePathEntry in typeAnnotation.typePathEntries) {
                    append(typePathEntry.typePathKind)
                    append(", argument index ").append(typePathEntry.typeArgumentIndex)
                }
            }.toString()
        }.visibilityPredicate { typeAnnotation ->
            typeAnnotation.typePathEntries.isNotEmpty()
        }
    }

    private fun handleLink(description: String) {
        val type = description[0]
        val index = Integer.parseInt(description.substring(1))
        when (type) {
            'E' -> handleExceptionsLink(index)
            'L' -> handleLocalVarLink(index)
            'I' -> handleInterfaceLink(index)
            else -> throw IllegalArgumentException("Invalid link type " + type)
        }
    }

    private fun handleExceptionsLink(index: Int) {
        val path = findParentNode(CodeAttribute::class.java, tree.selectionPath)
        selectPath(path)
        val detailPane = services.browserComponent.detailPane.currentDetailPane as AttributeDetailPane
        val codeAttributeDetailPane = detailPane.getDetailPane(CodeAttribute::class.java) as CodeAttributeDetailPane
        codeAttributeDetailPane.selectExceptionTableDetailPane()
        codeAttributeDetailPane.exceptionTableDetailPane.selectIndex(index)
    }

    private fun handleLocalVarLink(index: Int) {
        val (path, attribute) = findAttributeViaParent(CodeAttribute::class.java, LocalVariableTableAttribute::class.java)
        handleListLink(getLinkIndex(index, attribute), path, attribute)
    }

    private fun <T : AttributeInfo> findAttributeViaParent(parentClass: Class<out Structure>, attributeClass: Class<T>): Pair<TreePath, T> {
        val path = findAttributeChildNode(findParentNode(parentClass, tree.selectionPath), attributeClass)
        return Pair(path, attributeClass.cast((path.lastPathComponent as BrowserTreeNode).element))
    }

    private fun getLinkIndex(index: Int, attribute: LocalVariableTableAttribute): Int {
        val localVariableTable = attribute.localVariableEntries
        localVariableTable.forEachIndexed { i, localVariableEntry ->
            if (localVariableEntry.index == index) {
                return i
            }
        }
        throw IllegalArgumentException("index $index not found in local variable table")
    }

    private fun handleInterfaceLink(index: Int) {
        val interfacesPath = services.browserComponent.treePane.getPathForCategory(NodeType.INTERFACE)
        val interfacesNode = interfacesPath.lastPathComponent as BrowserTreeNode
        if (index >= interfacesNode.childCount) {
            throw IllegalArgumentException("Invalid interface index " + index)
        }
        val path = interfacesPath.pathByAddingChild(interfacesNode.getChildAt(index))
        selectPath(path)
    }

    private fun handleListLink(index: Int, path: TreePath, attribute: AttributeInfo) {
        selectPath(path)
        val detailPane = services.browserComponent.detailPane.currentDetailPane as AttributeDetailPane
        (detailPane.getDetailPane(attribute::class.java) as TableDetailPane).selectIndex(index)
    }

    private fun selectPath(path: TreePath) {
        tree.apply {
            selectionPath = path
            scrollPathToVisible(path)
        }
    }

    private val tree: JTree
        get() = services.browserComponent.treePane.tree

    private fun findAttributeChildNode(path: TreePath, attributeClass: Class<out AttributeInfo>): TreePath {
        val node = path.lastPathComponent as BrowserTreeNode
        node.children().iterator().forEach { child ->
            val attributeNode = child as BrowserTreeNode
            if (attributeNode.element?.let { it::class.java } == attributeClass) {
                return path.pathByAddingChild(attributeNode)
            }
        }
        throw RuntimeException("No attribute node for class $attributeClass found")
    }

    private fun findParentNode(elementClass: Class<out Any>, path: TreePath?): TreePath {
        if (path == null) {
            throw RuntimeException("No parent node with element class $elementClass found")
        }
        val node = path.lastPathComponent as BrowserTreeNode
        if (node.element?.let { it::class.java } == elementClass) {
            return path
        } else {
            return findParentNode(elementClass, path.parentPath)
        }
    }

}
